src/url.rs

```diff
@@ -190,6 +198,26 @@ impl PyUrl {
         }
         cls.call1((url,))
     }
+
+    #[pyo3(signature=(path, trailing_slash=true))]
+    pub fn join(&self, path: &str, trailing_slash: bool) -> PyResult<Self> {
+        let mut new_url = self
+            .lib_url
+            .join(path)
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+        if !trailing_slash || new_url.query().is_some() || new_url.fragment().is_some() || new_url.cannot_be_a_base() {
+            return Ok(PyUrl::new(new_url));
+        }
+
+        new_url
+            .path_segments_mut()
+            .map_err(|()| PyValueError::new_err("Url cannot be a base"))?
+            .pop_if_empty()
+            .push("");
```

Comment:
I think the "cannot be a base" restriction seems unnecessary. How about this alternative:

```suggestion
        if trailing_slash && || !new_url.path().ends_with("/") && !(new_url.query().is_some() || new_url.fragment().is_some()) {
            match new_url.path_segments_mut() {
                Ok(segments) => segments.pop_if_empty().push(),
                Err(()) {
                    let mut new_path = new_url.path().to_string();
                    new_path.push('/');
                    new_url.set_path(&new_path);
                }
            }
        }
```

Reply:
Yes, I agree this alternative looks better.